### DISCUSSION TITLE
Shared TCP Gateway

### DISCUSSION THREAD

**Discussion by deerlord** [markqvist/Reticulum#145] [Nomad Network]
Currently attempting to set up an instance that can serve as a central server for a set of clients. I have `enable_transport = Yes` in my config, and the following added to my config

  [[TCP Server Interface]]
    mode = gateway
    type = TCPServerInterface
    interface_enabled = True
    listen_ip = 0.0.0.0
    listen_port = 4242
    
   
   Clients have the appropriate TPCClientInterface defined for this server in their .reticulum/config files. Announcements from clients seem to generate traffic on the TCPServer interface, but these announcements are never received on other clients. Is there more to the config that is needed to ensure this type of setup works?

**Comment by markqvist** [comment]
No, this should definitely work as is. Something is probably wrong here, I am investigating it now.

**Comment by markqvist** [comment]
I tried replicating this problem with a setup consisting of one Reticulum instance with a TCP server interface, and two clients with a TCP client interface connecting to the server. Transport was enabled on the instance with the TCP server interface. Using various programs to test with, announces were passed between all Reticulum instances of the network.

The configs I used follows below. Please note, that since I ran all these instances on the same system, I changed the `shared_instance_port` and `instance_control_port` for each instance, so each instance would run in complete isolation. If you are running the Reticulum instances on *separate* systems, you don't need to do this, and can just leave that at the default. I also increased log level to 7, to help see any potential problems.

**Reticulum config on instance with TCP server interface**
```
[reticulum]
  enable_transport = True
  share_instance = Yes
  shared_instance_port = 37438
  instance_control_port = 37439

[logging]
  loglevel = 7

[interfaces]
    [[TCP Server Interface]]
    type = TCPServerInterface
    interface_enabled = True

    listen_ip = 0.0.0.0
    listen_port = 4242
```


**Reticulum config on first instance with TCP client interface**
```
[reticulum]
  enable_transport = False
  share_instance = Yes
  shared_instance_port = 37448
  instance_control_port = 37449

[logging]
  loglevel = 7

[interfaces]

  [[TCP Client Interface]]
    type = TCPClientInterface
    interface_enabled = True
    target_host = 127.0.0.1
    target_port = 4242
```


**Reticulum config on second instance with TCP client interface**
```
[reticulum]
  enable_transport = False
  share_instance = Yes
  shared_instance_port = 37458
  instance_control_port = 37459

[logging]
  loglevel = 7

[interfaces]

  [[TCP Client Interface]]
    type = TCPClientInterface
    interface_enabled = True
    target_host = 127.0.0.1
    target_port = 4242
```

With this setup, testing with for example the `nomadnet` program, announces propagate correctly between all instances, and can be seen both in the Reticulum log files and in the user-facing programs.

I was unfortunately completely unable to replicate this issue.

Can you please post the exact Reticulum configs that you used, and also let me know how you are testing the announce propagation?

**Reply by deerlord** [reply]
I was able to get this working, however I am not sure my LXMF Propagation Node is working correctly on the server. I can see both announces ("delivery destination" and "propagation node") on the server, as well as on the client, in the related log files. However, the NomadNet UI only shows the peer announce under Network -> Announce Stream, it does not show the node announce.

When connected to the test nets, I see both peer and node announces show up in the Announce Stream. This leads me to believe something is not correct with the node I am running, however there is no indication in the logs of what the issue might be. Below is my lxmf conf on the server

```
[propagation]
enable_node = yes
announce_interval = 1
announce_at_start = yes
autopeer = yes
autopeer_maxdepth = 4
message_storage_limit = 2000
auth_required = no
[lxmf]
announce_at_start = yes
announce_interval = 1
display_name = TestNode
[logging]
loglevel = 7
```

**Reply by markqvist** [reply]
What version of NomadNet are you running? And, from the config file, I take it you are using `lxmd` to host the propagation node, right?

**Comment by deerlord** [comment]
On the client end, which only runs NomadNet (reticulum and lxmd versions provided for explicitness)
```
nomadnet --version
Nomad Network Client 0.2.5
rnsd --version
rnsd 0.3.18
lxmd --version
lxmd 0.2.2
```

and the node, which is running `rnsd` and `lxmd`
```
rnsd --version
rnsd 0.4.0
lxmd --version
lxmd 0.2.3
```

**Reply by markqvist** [reply]
Ok, great! Try updating NomadNet to 0.2.7: `pip install nomadnet --upgrade`

Support for displaying the standalone `lxmd`-powered nodes was only just added in version 0.2.6, so that is probably just the problem.

**Reply by deerlord** [reply]
That did seem to work, I do see the announces and more information about the node. It does not seem message propagation is actually working, but I need to do some more debugging around that issue. The nomad client at least recognizes the node and logs that it will propagate messages to it.

If there's no known issues that might impact this, I will go ahead and close this discussion.

**Reply by markqvist** [reply]
What parts are not working? Can NomadNet / Sideband deliver messages via the `lxmd` node? Can they sync messages from it? Or is none of that working at all? Or is it only propagation *between* nodes that are not working?
